### PR TITLE
watchdog: open /dev/watchdog0 only if it exists

### LINKED ISSUE #24661: watchdog support doesn't support legacy watchdog interface /dev/watchdog anymore, but ipmi_watchdog driver only supports that
### systemd version the issue has been seen with

250.4

### Used distribution

NixOS

### Linux kernel version used

5.15.67

### CPU architectures issue was seen on

x86_64

### Component

_No response_

### Expected behaviour you didn't see

I configured the explicit device:

```
[root@wendy00:/etc/nixos]# grep "WatchdogDevice" /etc/systemd/system.conf
WatchdogDevice=/dev/watchdog
```

However, system keeps trying /dev/watchdog0:

```
Sep 13 19:35:44 wendy00 systemd[1]: Failed to open watchdog device /dev/watchdog0, ignoring: No such file or directory
```

When symlinking /dev/watchdog0 to /dev/watchdog (this is an IPMI watchdog) it picks up the device properly.

However, I'm also confused because the documented option is not visible in any "show" inquiries:

```
[root@wendy00:/etc/nixos]# systemctl show --property=WatchdogDevice

[root@wendy00:/etc/nixos]# echo $?
0
```

I haven't found any other errors that would indicate me misspelling the property name.

### Unexpected behaviour you saw

_No response_

### Steps to reproduce the problem

_No response_

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
Sep 13 19:35:44 wendy00 systemd[1]: Failed to open watchdog device /dev/watchdog0, ignoring: No such file or directory
```

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #24661.

### CODE REVIEW

**Comment by bluca** [comment]
@mrc0mmand vagrant looks sad today:

```
23:06:24 cp: cannot stat '/root/systemd/vagrant-*': No such file or directory
23:06:24 cp: cannot stat '/tmp/vagrant-*-console.log': No such file or directory
23:06:24 2022-09-13 22:06:24,474 [agent-control/fetch_artifacts] INFO: Fetching artifacts from node n27-5-5.pool.ci.centos.org: (remote: ~/vagrant-logs*, local: ./artifacts_wqqfuldt)
23:06:24 2022-09-13 22:06:24,474 [agent-control/execute_local_command] INFO: Executing a LOCAL command: /usr/bin/scp -r -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root@n27-5-5.pool.ci.centos.org:~/vagrant-logs* ./artifacts_wqqfuldt
23:06:24 2022-09-13 22:06:24,643 [agent-control/main] ERROR: Execution failed
23:06:24 Traceback (most recent call last):
23:06:24   File "./agent-control.py", line 403, in main
23:06:24     ac.execute_remote_command(command, artifacts_dir="~/vagrant-logs*")
23:06:24   File "./agent-control.py", line 164, in execute_remote_command
23:06:24     raise RuntimeError("Remote command exited with an unexpected return code "
23:06:24 RuntimeError: Remote command exited with an unexpected return code (got: 1, expected: 0), bailing out
```

**Comment by mrc0mmand** [comment]
> @mrc0mmand vagrant looks sad today:
> 
> ```
> 23:06:24 cp: cannot stat '/root/systemd/vagrant-*': No such file or directory
> 23:06:24 cp: cannot stat '/tmp/vagrant-*-console.log': No such file or directory
> 23:06:24 2022-09-13 22:06:24,474 [agent-control/fetch_artifacts] INFO: Fetching artifacts from node n27-5-5.pool.ci.centos.org: (remote: ~/vagrant-logs*, local: ./artifacts_wqqfuldt)
> 23:06:24 2022-09-13 22:06:24,474 [agent-control/execute_local_command] INFO: Executing a LOCAL command: /usr/bin/scp -r -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root@n27-5-5.pool.ci.centos.org:~/vagrant-logs* ./artifacts_wqqfuldt
> 23:06:24 2022-09-13 22:06:24,643 [agent-control/main] ERROR: Execution failed
> 23:06:24 Traceback (most recent call last):
> 23:06:24   File "./agent-control.py", line 403, in main
> 23:06:24     ac.execute_remote_command(command, artifacts_dir="~/vagrant-logs*")
> 23:06:24   File "./agent-control.py", line 164, in execute_remote_command
> 23:06:24     raise RuntimeError("Remote command exited with an unexpected return code "
> 23:06:24 RuntimeError: Remote command exited with an unexpected return code (got: 1, expected: 0), bailing out
> ```

Whoops, I forgot to reschedule this one. I have this issue on my TODO list with a possible solution, hopefully will have a fix ready soon, thanks for the ping.

**Comment by bluca** [comment]
/dev/watchdog0 should exist on all supported kernel versions, let's wait for the reporter to provide more information

File: src/shared/watchdog.c [hunk 1]
```diff
@@ -308,16 +307,26 @@ static int open_watchdog(void) {
...
+        }
 
-        watchdog_fd = open(watchdog_device, O_WRONLY|O_CLOEXEC);
-        if (watchdog_fd < 0)
-                return log_debug_errno(errno, "Failed to open watchdog device %s, ignoring: %m", watchdog_device);
+        if (watchdog_fd < 0) {
+                watchdog_fd = open(watchdog_device, O_WRONLY|O_CLOEXEC);
+                if (watchdog_fd < 0)
+                        return log_debug_errno(errno, "Failed to open watchdog device %s: %m", watchdog_device);
+        }
```
Comment (poettering) [inline src/shared/watchdog.c:342]: what about this:

```c
char **try_order;

if (!watchdog_device || PATH_IN_SET(watchdog_device, "/dev/watchdog0", "/dev/watchdog"))
        try_order = STRV_MAKE("/dev/watchdog0", "/dev/watchdog");
else
        try_order = STRV_MAKE(watchdog_device);

STRV_FOREACH(wd, try_order) {
        watchdog_fd = open(wd, O_WRONLY|O_CLOEXEC);
        if (watchdog >= 0) {
                r = free_and_strdup(&watchdog_device, wd);
                if (r < 0)
                        return r;
                break; 
        } 
        
        if (errno != ENOENT)
                return log_debug_errno(errno, "Failed to open watchdog device %s: %m", wd);
}

if (watchdog_fd < 0)
       return log_debug_errno(SYNTHETIC_ERRNO(ENOENT), "Failed to open watchdog device %s: %m", watchdog_device ?: "auto");
```
or so?

i.e. always prefer the new-style devices over the old ones, regardless what precisely you specify for the first device.

Comment (yuwata) [reply src/shared/watchdog.c:342]: I cannot find big difference between the current and suggested code. and unfortunately `STRV_MAKE` in if block cannot used outside (or course, we can by using ternary op, though).

**Comment by yuwata** [comment]
Applied the suggestion. PTAL.

**Review by bluca — APPROVED** [review]

File: src/shared/watchdog.c [hunk 2]
```diff
@@ -324,16 +324,25 @@ static int open_watchdog(void) {
          * has the benefit that we can easily find the matching directory in sysfs from it, as the relevant
          * sysfs attributes can only be found via /sys/dev/char/<major>:<minor> if the new-style device
          * major/minor is used, not the old-style. */
-        fn = !watchdog_device || path_equal(watchdog_device, "/dev/watchdog") ?
-                "/dev/watchdog0" : watchdog_device;
+        try_order = !watchdog_device || path_equal(watchdog_device, "/dev/watchdog") ?
```
Comment (poettering) [inline src/shared/watchdog.c]: why not `PATH_IN_SET(watchdog, "/dev/watchdog", "/dev/watchdog0")` here?

i.e. I'd always consider these devices equivalent, and given they are equivalent we want to use the better one, always, but if the better one doesn't exist the worse one.

Comment (yuwata) [reply src/shared/watchdog.c]: Simply I forgot to apply that from your suggestion :-)

**Pull request merged**